core/log/batch_logger.cpp

```diff
@@ -10,9 +10,66 @@
 #include <ginkgo/core/base/math.hpp>
 
 
+#include "core/base/workspace_aliases.hpp"
+
+
 namespace gko {
 namespace batch {
 namespace log {
+namespace detail {
+
+
+template <typename ValueType>
+log_data<ValueType>::log_data(std::shared_ptr<const Executor> exec,
+                              size_type num_batch_items)
+    : res_norms(exec), iter_counts(exec)
+{
+    if (num_batch_items > 0) {
+        iter_counts.resize_and_reset(num_batch_items);
+        res_norms.resize_and_reset(num_batch_items);
+    } else {
+        GKO_INVALID_STATE("Invalid num batch items passed in");
+    }
+}
+
+
+template <typename ValueType>
+log_data<ValueType>::log_data(std::shared_ptr<const Executor> exec,
+                              size_type num_batch_items,
+                              array<unsigned char>& workspace)
+    : res_norms(exec), iter_counts(exec)
+{
+    const size_type reqd_workspace_size = num_batch_items * 32;
```

Comment:
If I understand it correctly, the mapping is to ensure the memory address is divisible by the size of data type.
It's also why I think it was enough by viewing the larger data type first.
Assume the workspace data is not aliased from the beginning, which should not be the case if the workspace is the alone allocation, and then the actual memory will need a little more than reqd_workspace_size by cutting the first unaligned part. If the reqd_workspace_size is the same as the workspace allocation, it does not help when the workspace is not aligned from the beginning.

Reply:
okay, I may misunderstand the `reqd_workspace_size` meaning. I thought it is required workspace size. That's why I thought it will be failed when it is not aligned. The actual size is larger than required size == workspace size
However, it is just the size of workspace pointer in `create_workspace_aliases`.
It will be required workspace size (by function) when the aliases phase is failed.